### PR TITLE
[Feature] Add Typo tolerance

### LINKED ISSUE #152: Add support to the typo tolerance customization
⚠️ This issue is generated, it means the examples and the namings do not necessarily correspond to the language of this repository.
Also, if you are a maintainer, feel free to add any clarification and instruction about this issue.

Related to:
- parent issue: https://github.com/meilisearch/integration-guides/issues/190
- specification: https://github.com/meilisearch/specifications/pull/117

----

Add a `typoTolerance` index settings resource to manage customization of the typo tolerance feature at the index level.

- Expose `typoTolerance` object on `/indexes/:index_uid/settings` endpoints.
- Add `GET/POST/DELETE` - `/indexes/:index_uid/settings/typo-tolerance` endpoints.

The `typoTolerance` object definition is made of the following properties:
  - `enabled` type (`boolean`)
  - `disableOnAttributes` type (`array[string]`)
  - `disableOnWords` type (`array[string]`)
  - `minWordSizeForTypos` type (`object`)
    - `oneTypo` type (`int`)
    - `twoTypos` type (`int`)

More information about what each field does could be [checked in the spec.](https://github.com/meilisearch/specifications/pull/117)

Check the other customization methods for reference like: `synonyms`, `searchableAttributes`, `filterableAttributes`.

TODO:
- [ ] Add the new methods to customize `typo-tolerance`
  - [ ] GET `/indexes/:index_uid/settings/typo-tolerance`
  - [ ] POST `/indexes/:index_uid/settings/typo-tolerance`
  - [ ] DELETE `/indexes/:index_uid/settings/typo-tolerance`
- [ ] Add tests

### PR DESCRIPTION
**Pull request by ahmednfwela**
Fixes #152
Depends on #262

- [x] Tests Added

### CODE REVIEW

**Comment by brunoocasali** [comment]
Can you fix the conflicts here?

**Comment by ahmednfwela** [comment]
Conflicts fixed and rebased on #262

File: lib/src/index_settings.dart [hunk 1]
```diff
@@ -45,22 +51,28 @@ class IndexSettings {
...
+          (map['filterableAttributes'] as List?)?.cast<String>(),
+      distinctAttribute: (map['distinctAttribute'] as String?),
+      searchableAttributes:
+          (map['searchableAttributes'] as List?)?.cast<String>(),
+      displayedAttributes:
+          (map['displayedAttributes'] as List?)?.cast<String>(),
+      sortableAttributes: (map['sortableAttributes'] as List?)?.cast<String>(),
+      typoTolerance: typoTolerance is Map<String, Object?>
+          ? TypoToleranceSettings.fromMap(typoTolerance)
+          : null,
```
[the remark this answers is not in the document]
Comment (ahmednfwela) [reply lib/src/index_settings.dart:75]: I ended up making the others use the same pattern for type safety

**Comment by ahmednfwela** [comment]
I moved all the index settings into a settings folder and added a barrel file to export them cleanly

**Review by brunoocasali — APPROVED** [review]
Thanks @ahmednfwela 🎉 

bors merge

**Pull request merged**